pgautoupgrade-postupgrade.sh

```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env bash
+
+set -e
+
+if [ $# -ne 2 ]; then
+	echo "Required number of arguments not passed to post upgrade script.  2 expected, $# received"
+	exit 1
+fi
+
+PGDATA=$1
+PGAUTO_ONESHOT=$2
+
+# Wait for PostgreSQL to start and become available
+COUNT=0
+RUNNING=1
+while [ $RUNNING -ne 0 ] && [ $COUNT -le 20 ]; do
+	# Check if PostgreSQL is running yet
+	echo "------ Checking if PostgreSQL is running, loop count ${COUNT} ------"
+	set +e
+	pg_isready -q
+	RUNNING=$?
+	set -e
+
+	if [ $RUNNING -eq 0 ]; then
+		echo "PostgreSQL is running.  Post upgrade tasks will start shortly"
+	else
+		echo "PostgreSQL is not yet running, lets wait then try again..."
+		sleep 3
+	fi
+
+	COUNT=$((COUNT+1))
+done
+
+if [ $RUNNING -ne 0 ]; then
+	echo "PostgreSQL did not start before timeout expired"
+	exit 2
+fi
+
+# Get the list of databases in the database cluster
+DB_LIST=$(echo 'SELECT datname FROM pg_catalog.pg_database WHERE datistemplate IS FALSE' | psql -1t --csv postgres)
+
+# Update query planner statistics
+echo "----------------------------"
+echo "Updating query planner stats"
+echo "----------------------------"
+
+for DATABASE in ${DB_LIST}; do
+	echo "VACUUM (ANALYZE, VERBOSE, INDEX_CLEANUP FALSE)" | psql -t --csv "${DATABASE}"
+done
+
+echo "-------------------------------------"
+echo "Finished updating query planner stats"
+echo "-------------------------------------"
+
+# Reindex the databases
+echo "------------------------"
+echo "Reindexing the databases"
+echo "------------------------"
+
+# For each database, reindex it
+for DATABASE in ${DB_LIST}; do
+	echo "-------------------------------"
+	echo "Starting reindex of ${DATABASE}"
+	echo "-------------------------------"
+
+	echo 'REINDEX DATABASE CONCURRENTLY' | psql -t --csv "${DATABASE}"
+
+	echo "-------------------------------"
+	echo "Finished reindex of ${DATABASE}"
+	echo "-------------------------------"
+done
+
+echo "-------------------------------"
+echo "End of reindexing the databases"
+echo "-------------------------------"
+
+# If "one shot" mode was requested, then shut down PostgreSQL
+if [ "x${PGAUTO_ONESHOT}" = "xyes" ]; then
```

Comment:
not sure if we should also test that the one shot mode works correctly with our `test.sh`. because there are a couple of "special" things that had to be implemented.

Reply:
Sounds like a good idea.  Will need to think on it. :smile:

Reply:
For this, I'm thinking the only real testing we can do is (somehow) to check whether the one shot mode does indeed shut down automatically after starting the container.

Probably need to do some kind of timeout based check whether the service is still running after launching, and also check that the $PGDATA/PG_VERSION file is what we expect after the run.  Will experiment with it a bit later. :smile:

Reply:
Just added basic testing of the one shot mode too.  Instead of using a timeout (or similar) it just uses the far more simple `docker compose run` approach then tests the `PG_VERSION` file in the data directory afterwards.

That'll probably do for now. :smile: